### PR TITLE
Add Metadata to codec

### PR DESCRIPTION
**Pull request by soya3129**
Signed-off-by: Yang Song <yasong@yasong00.cam.corp.google.com>

*Description*: Add MetadataEncoder and MetadataDecoder to codec. I also corrected some of the design decisions in MetadataEnocder and MetadataDecoder so that they are more suitable for codec. Those corrections includes:  1. let StreamImpl owns the decoded metadata, not the decoder. 2. Remove stream_id field in decoder and encoder since stream id is stored in StreamImpl. 3. Remove set/getMaxMetadataSize(), because nghttp2 uses default max frame size.

This is the second step towards supporting metadata in envoy. The next step would be add filters to add/delete/modify metadata.

*Risk Level*: Low. Not used.
*Testing*: Unit test.
*Docs Changes*: 
*Release Notes*:
#2394

### CODE REVIEW

File: include/envoy/http/codec.h [hunk 1]
```diff
@@ -52,6 +53,12 @@ class StreamEncoder {
    * @return Stream& the backing stream.
    */
   virtual Stream& getStream() PURE;
+
+  /**
+   * Encode METADATA.
+   * @param metadata_map is the METADATA to encode.
+   */
+  virtual void encodeMetadata(const MetadataMap& metadata_map) PURE;
```
Comment (alyssawilk) [inline include/envoy/http/codec.h]: I think this is a good PR to start landing docs with, even if we have them tagged as work in progress.
Let's talk about when you can encodeMetadata, how often you can encode it, when it will be passed up to filters (we can edit the docs if they change when that lands) etc.

File: source/common/http/conn_manager_impl.h [hunk 2]
```diff
@@ -303,6 +303,7 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     void decodeHeaders(HeaderMapPtr&& headers, bool end_stream) override;
     void decodeData(Buffer::Instance& data, bool end_stream) override;
     void decodeTrailers(HeaderMapPtr&& trailers) override;
+    void decodeMetadata(MetadataMap&) override { NOT_REACHED_GCOVR_EXCL_LINE; }
```
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.h]: Do we think there should be anything done if an HTTP/1.1 codec is asked to encode metadata?  If nothing else maybe error reporting stats so folks know their metadata is being swallowed?

File: source/common/http/http2/codec_impl.cc [hunk 3]
```diff
@@ -48,8 +48,8 @@ template <typename T> static T* remove_const(const void* object) {
 }
 
 ConnectionImpl::StreamImpl::StreamImpl(ConnectionImpl& parent, uint32_t buffer_limit)
-    : parent_(parent), headers_(new HeaderMapImpl()), local_end_stream_sent_(false),
-      remote_end_stream_(false), data_deferred_(false),
+    : parent_(parent), headers_(new HeaderMapImpl()), metadata_map_(new MetadataMap()),
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc]: What do we think about doing lazy creation here?  I think it's worth the complexity to not create a map for the vast majority of users who don't need this functionality.

Also if there's one metadata_map_ (at a time) per stream and one decoder per stream I'm not sure I follow why you moved ownership out of the decoder.

File: source/common/http/http2/codec_impl.cc [hunk 4]
```diff
@@ -543,6 +577,57 @@ int ConnectionImpl::onStreamClose(int32_t stream_id, uint32_t error_code) {
...
+  ASSERT(len >= size_to_copy);
+
+  Buffer::OwnedImpl& p = reinterpret_cast<Buffer::OwnedImpl&>(encoder.payload());
+  p.copyOut(0, size_to_copy, buf);
+
+  // Releases the payload that has been copied to nghttp2.
+  encoder.releasePayload(size_to_copy);
+
+  // Keep submitting extension frames if there is payload left in the encoder.
+  if (encoder.hasNextFrame()) {
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc]: so when the caller calls encodeMetadata with a large frame, encodeMetadata calls submitMetadata, submitMetadata results in nghttp2 calling packMetadata, which may result in another submitMetadata?

Any chance we can remove the recursion by moving where we check for hasNextFrame and calling submitMetadata multiple times instead?

File: source/common/router/router.h [hunk 5]
```diff
@@ -288,6 +288,7 @@ class Filter : Logger::Loggable<Logger::Id::router>,
     void decodeHeaders(Http::HeaderMapPtr&& headers, bool end_stream) override;
     void decodeData(Buffer::Instance& data, bool end_stream) override;
     void decodeTrailers(Http::HeaderMapPtr&& trailers) override;
+    void decodeMetadata(Http::MetadataMap&) override { NOT_REACHED_GCOVR_EXCL_LINE; }
```
Comment (alyssawilk) [inline source/common/router/router.h]: I think we're going to need to implement something here.  Maybe a TODO instead of marking it as not reached?

File: source/common/upstream/health_checker_impl.h [hunk 6]
```diff
@@ -280,6 +281,7 @@ class GrpcHealthCheckerImpl : public HealthCheckerImplBase {
     void decodeHeaders(Http::HeaderMapPtr&& headers, bool end_stream) override;
     void decodeData(Buffer::Instance&, bool end_stream) override;
     void decodeTrailers(Http::HeaderMapPtr&&) override;
+    void decodeMetadata(Http::MetadataMap&) override {}
```
Comment (alyssawilk) [inline source/common/upstream/health_checker_impl.h]: These ones we're probably never going to really use, but we could unit test the functions to avoid a coverage decrease

File: include/envoy/http/metadata_interface.h [hunk 7]
```diff
@@ -22,10 +21,13 @@ namespace Http2 {
 const uint8_t METADATA_FRAME_TYPE = 0x4d;
 const uint8_t END_METADATA_FLAG = 0x4;
 
+// NGHTTP2_MAX_PAYLOADLEN in nghttp2.
+const uint64_t METADATA_MAX_PAYLOAD_SIZE = 16384;
```
Comment (alyssawilk) [inline include/envoy/http/metadata_interface.h]: As a sanity check, if nghttp2 changed their NGHTTP2_MAX_PAYLOADLEN to be lower, they'd reject large data chunks and your large metadata frame test would fail when we imported the change, right?

File: include/envoy/http/codec.h (hunk 1 above)
Comment (soya3129) [reply include/envoy/http/codec.h]: Still working on the doc.

File: source/common/http/conn_manager_impl.h (hunk 2 above)
Comment (soya3129) [reply source/common/http/conn_manager_impl.h]: Good call! I added an error log in Http1::StreamEncoderImpl::encodeHeader().

File: source/common/http/http2/codec_impl.cc (hunk 3 above)
Comment (soya3129) [reply source/common/http/http2/codec_impl.cc]: Right! Moved metadata_map_ back to decoder. I thought it would be consistent to let StreamImpl hold metadata since it holds headers. But it is definitely not necessary.

File: source/common/http/http2/codec_impl.cc (hunk 4 above)
Comment (soya3129) [reply source/common/http/http2/codec_impl.cc]: You are right. The old code let packMetadata call submitMetadata until there is no metadata left. I can see it is very hard to understand. Changed according to the suggestion.

File: source/common/router/router.h (hunk 5 above)
Comment (soya3129) [reply source/common/router/router.h]: Done.

File: source/common/upstream/health_checker_impl.h (hunk 6 above)
Comment (soya3129) [reply source/common/upstream/health_checker_impl.h]: Good idea! Test added.

File: include/envoy/http/metadata_interface.h (hunk 7 above)
Comment (soya3129) [reply include/envoy/http/metadata_interface.h]: Great question! I increase METADATA_MAX_PAYLOAD_SIZE to 20000, and the test fails at:
source/common/http/http2/codec_impl.cc:623
ASSERT(len >= size_to_copy);

Added a comment to check NGHTTP2_MAX_PAYLOADLEN in case of the assert failure.

**Comment by soya3129** [comment]
Thanks for the review! I am still working on the doc part. Will update the PR when the doc is ready.

I changed two if conditions to assert since the if condition should not happen, and it increases the test coverage. I am having trouble making tests hit the if condition in onMetadataReceived():
if (!stream) { 
 return 0
}
It can't be changed to ASSERT in case envoy receives metadata with invalid stream id. I also notice the existing function saveHeader() function has the same problem. I will leave it for now if there is no objection.

File: source/docs/h2_metadata.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+Envoy can consume or add metadata. But for one request (or response), we expect at most one of the
+two operations to be performed.
+TODO(soya3129): check if this limitation is needed after the implementation is
+complete.
+
+### Envoy metadata handling
+
+Envoy provides the functionality to proxy, process and add metadatas.
+
+## Proxy metadata
```
Comment (alyssawilk) [inline source/docs/h2_metadata.md]: Let's have the headers consistent
Proxying meadata
Consuming metadata
Inserting metadata

File: source/docs/h2_metadata.md [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+If not specified, all the metadata received by Envoy is proxied to the next hop
+unmodified. Note that, we do not guarantee the same frame order will be preserved from
+hop by hop. That is, metadata from upstream at the beginning of a stream can be
+received by the downstream at the end of the stream.
+
+## How to consume metadata
+
+(To be implemented)
+
+If Envoy needs to take actions when a metadata is received, users should
```
Comment (alyssawilk) [inline source/docs/h2_metadata.md]: s/a metadata/metadata/
or "a metadata frame"?

File: source/docs/h2_metadata.md [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+StreamDecodeFilter should be created. The interface to override is
+
+FilterTrailersStatus StreamDecoderFilter::decodeMetadata(MetadataMap& metadatas);
+
+The metadata passed in is a map of the metadatas associated with the request stream. After metadatas
+have been parsed, the filter can choose to remove metadata from the map, or keep
+it untouched.
+
+If Envoy needs to parse a metadata sent on a response from upstream to downstream, a
+StreamEncodeFilter should be created. The interface to override is
```
Comment (alyssawilk) [inline source/docs/h2_metadata.md]: Encoder

File: source/docs/h2_metadata.md [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+END\_METADATA (0x4).
+
+If the flag is set, it indicates that this frame ends a metadata
+payload.
+
+The METADATA frame payload is not subject to HTTP/2 flow control, but the size
+of the payload is bounded by the maximum frame size negotiated in SETTINGS.
+There are no restrictions on the set of octets that may be used in keys or values.
+TODO(soya3129): decide if we allow METADATA frame to terminate a stream.
```
Comment (alyssawilk) [inline source/docs/h2_metadata.md]: I'd say no for now.  Folks can send 0 length data frame.

File: source/docs/h2_metadata.md [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+Envoy provides the functionality to proxy, process and add metadatas.
+
+## Proxy metadata
+
+(To be implemented)
+
+If not specified, all the metadata received by Envoy is proxied to the next hop
+unmodified. Note that, we do not guarantee the same frame order will be preserved from
+hop by hop. That is, metadata from upstream at the beginning of a stream can be
+received by the downstream at the end of the stream.
```
Comment (alyssawilk) [inline source/docs/h2_metadata.md]: So I agree that we don't guarantee anything with respect to body order (and doccing that up is an excellent idea!) but I think we can and probably should make sure we don't mess with end stream.  For example if we forward end-stream ahead of metadata in a response headed downstream that stream may be torn down before the metadata is received.  I think we have to make sure that if we've read metadata we defer end stream until metadata is flushed through the filter chain.  What do you think?

**Comment by soya3129** [comment]
Please hold the review until I fix the check failure. Sorry about the noisy notification! I will address the check failure and the documentation comments together.

File: include/envoy/http/codec.h [hunk 13]
```diff
@@ -86,6 +93,12 @@ class StreamDecoder {
    * @param trailers supplies the decoded trailers.
    */
   virtual void decodeTrailers(HeaderMapPtr&& trailers) PURE;
+
+  /**
+   * Called with decoded METADATA.
+   * @param decoded METADATA.
+   */
+  virtual void decodeMetadata(MetadataMap& metadata_map) PURE;
```
Comment (mattklein123) [inline include/envoy/http/codec.h]: Do we want to pass this as a unique_ptr with move semantics so that the callee can take ownership if they desire? I suspect this is what you want, much like the other header decode APIs.

File: source/common/http/http1/codec_impl.h [hunk 14]
```diff
@@ -37,6 +37,9 @@ class StreamEncoderImpl : public StreamEncoder,
   void encodeHeaders(const HeaderMap& headers, bool end_stream) override;
   void encodeData(Buffer::Instance& data, bool end_stream) override;
   void encodeTrailers(const HeaderMap& trailers) override;
+  void encodeMetadata(const MetadataMap&) override {
+    ENVOY_LOG(error, "HTTP/1.1 doesn't support METADATA. METADATA is ignored.");
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/http/http1/codec_impl.h]: Personally I would just do NOT_IMPLEMENTED and have this crash but I don't feel strongly about it.

File: source/common/http/http2/codec_impl.cc [hunk 15]
```diff
@@ -130,6 +130,27 @@ void ConnectionImpl::StreamImpl::encodeTrailers(const HeaderMap& trailers) {
     parent_.sendPendingFrames();
   }
 }
+
+void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMap& metadata_map) {
+  if (!parent_.allow_metadata_) {
+    ENVOY_CONN_LOG(error, "Connection doesn't allow metadata", parent_.connection_);
```
Comment (mattklein123) [inline source/common/http/http2/codec_impl.cc]: When do we envision this error happening? I'm mainly wondering if this should be a stat? Or should it just crash? Or just be a debug statement? I'm leery of error logs being useful to people.

File: source/common/http/http2/codec_impl.cc [hunk 16]
```diff
@@ -194,6 +215,15 @@ void ConnectionImpl::StreamImpl::submitTrailers(const HeaderMap& trailers) {
   ASSERT(rc == 0);
 }
 
+void ConnectionImpl::StreamImpl::submitMetadata() {
+  ASSERT(stream_id_ > -1);
```
Comment (mattklein123) [inline source/common/http/http2/codec_impl.cc]: We should never encode metadata on stream 0/connection either, right? I think this can be > 0?

File: source/common/http/http2/codec_impl.cc [hunk 17]
```diff
@@ -286,6 +316,26 @@ void ConnectionImpl::StreamImpl::resetStreamWorker(StreamResetReason reason) {
...
+  if (metadata_encoder_ == nullptr) {
+    metadata_encoder_ = std::make_unique<MetadataEncoder>();
+  }
+  return *metadata_encoder_;
+}
+
+MetadataDecoder& ConnectionImpl::StreamImpl::getMetadataDecoder() {
+  if (metadata_decoder_ == nullptr) {
+    auto cb =
+        std::bind(&ConnectionImpl::StreamImpl::onMetadataDecoded, this, std::placeholders::_1);
```
Comment (mattklein123) [inline source/common/http/http2/codec_impl.cc]: nit: I find bind hard to read, can you just pass an inline lambda / std::function?

File: source/common/http/http2/codec_impl.cc [hunk 18]
```diff
@@ -545,6 +595,50 @@ int ConnectionImpl::onStreamClose(int32_t stream_id, uint32_t error_code) {
...
+  StreamImpl* stream = getStream(stream_id);
+  ASSERT(stream != nullptr);
+
+  MetadataEncoder& encoder = stream->getMetadataEncoder();
+  const uint64_t size_to_copy = std::min(METADATA_MAX_PAYLOAD_SIZE, encoder.payload().length());
+  // nghttp2 guarantees len is at least 16KiB. If the check fails, please verify
+  // NGHTTP2_MAX_PAYLOADLEN is consistent with METADATA_MAX_PAYLOAD_SIZE.
+  ASSERT(len >= size_to_copy);
+
+  Buffer::OwnedImpl& p = reinterpret_cast<Buffer::OwnedImpl&>(encoder.payload());
```
Comment (mattklein123) [inline source/common/http/http2/codec_impl.cc]: Doing this type of cast is a bit of an anti-pattern. Why is this needed?

File: source/common/http/http2/codec_impl.h [hunk 19]
```diff
@@ -210,6 +221,8 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
         [this]() -> void { this->pendingSendBufferLowWatermark(); },
         [this]() -> void { this->pendingSendBufferHighWatermark(); }};
     HeaderMapPtr pending_trailers_;
+    std::unique_ptr<MetadataDecoder> metadata_decoder_{nullptr};
```
Comment (mattklein123) [inline source/common/http/http2/codec_impl.h]: nit: `{nullptr}` not needed. Default init will do this. Same next line.

**Comment by mattklein123** [comment]
Marking as waiting. 

/wait

File: include/envoy/http/codec.h (hunk 13 above)
Comment (soya3129) [reply include/envoy/http/codec.h]: Fixed as suggested. Thanks!

File: include/envoy/http/codec.h [hunk 20]
```diff
@@ -86,6 +93,12 @@ class StreamDecoder {
    * @param trailers supplies the decoded trailers.
    */
   virtual void decodeTrailers(HeaderMapPtr&& trailers) PURE;
+
+  /**
+   * Called with decoded METADATA.
+   * @param decoded METADATA.
+   */
+  virtual void decodeMetadata(std::unique_ptr<MetadataMap> metadata_map) PURE;
```
Comment (mattklein123) [inline include/envoy/http/codec.h]: I would pass as `std::unique_ptr<MetadataMap>&&` (see style doc)

Comment (soya3129) [reply include/envoy/http/codec.h]: Fixed.

**Comment by soya3129** [comment]
Thanks for the review!

File: source/common/http/http2/metadata_decoder.h [hunk 21]
```diff
@@ -58,18 +59,15 @@ class MetadataDecoder : Logger::Loggable<Logger::Id::http2> {
    */
   bool decodeMetadataPayloadUsingNghttp2(bool end_metadata);
 
+  // Metadata that is currently under decoding.
+  std::unique_ptr<MetadataMap> metadata_map_;
```
Comment (lizan) [inline source/common/http/http2/metadata_decoder.h]: nit: MetadataMapPtr

File: include/envoy/http/metadata_interface.h (hunk 7 above)
Comment (lizan) [reply include/envoy/http/metadata_interface.h]: Make this as constexpr (perhaps for all constants), and do `static_assert` against `NGHTTP2_MAX_PAYLOADLEN`? It will fail at compile time then.

File: source/common/http/http2/metadata_decoder.h [hunk 22]
```diff
@@ -19,7 +20,10 @@ class MetadataEncoderDecoderTest_VerifyEncoderDecoderOnMultipleMetadataMaps_Test
 // map of string key value pairs.
 class MetadataDecoder : Logger::Loggable<Logger::Id::http2> {
 public:
-  MetadataDecoder(uint64_t stream_id, MetadataCallback cb);
+  /**
+   * @param metadata_map is where to save the decoded metadata.
```
Comment (alyssawilk) [inline source/common/http/http2/metadata_decoder.h]: update comment?

File: source/common/http/http2/metadata_decoder.h (hunk 21 above)
Comment (alyssawilk) [reply source/common/http/http2/metadata_decoder.h]: +1, plus "is currently under decoding." -> "is currently being decoded."?

File: source/docs/h2_metadata.md [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+Envoy provides the functionality to proxy, process and add metadata.
+
+## Proxying metadata
+
+(To be implemented)
+
+If not specified, all the metadata received by Envoy is proxied to the next hop
+unmodified. Note that, we do not guarantee the same frame order will be preserved from
+hop by hop. That is, metadata from upstream at the beginning of a stream can be
+received by the downstream at the end of the stream.
```
Comment (alyssawilk) [inline source/docs/h2_metadata.md]: Worth commenting somewhere that empty maps will not be serialized out, i.e. if you clear the data it will functionally result in metadata being removed for the next hop?

File: source/common/http/http2/codec_impl.cc (hunk 15 above)
Comment (soya3129) [reply source/common/http/http2/codec_impl.cc]: Correct my comments above: we don't need to check allow_metadata_ not because of the reasons I stated. If allow_metadata_ is false, nghttp2 will not pass metadata to codec. We don't need to check it in envoy code anywhere. Sorry about the wrong information!

**Comment by soya3129** [comment]
Cc Biren the metadata change.
@birenroy
